File: python-dsa/queue/linkedqueue.py

"""
This module implements the Queue ADT using an underlying linkedlist
"""

class LinkedQueue(object):
    """Implementation of the queue using a singly linkedlist with head and tail pointer"""

    class _Node(object):
        """Internal class to represent linkedlist"""
        __slots__ = ['_next', '_element']
        def __init__(self, nextn, element):
            self._next = nextn
            self._element = element

    def __init__(self):
        self.__size = 0
        self.__head = None
        self.__tail = None

    def __len__(self):
        return self.__size

    def front(self):
        """Returns the element at the front of the queue"""
        if self.__size == 0:
            raise EmptyError('Queue is empty!')
        return self.__head._element

    def enqueue(self, element):
        """Add the element to the back of the queue"""
        node = self._Node(None, element)
        if self.__size == 0:
            self.__head = node
            self.__tail = node
            self.__size += 1
            return
        self.__tail._next = node
        self.__tail = node
        self.__size += 1

    def dequeue(self):
        """Remove and return the element at the front of the queue."""
        if self.__size == 0:
            raise EmptyError("The queue is empty!")
        val = self.__head._element
        self.__head = self.__head._next
        if not self.__head:
            self.__tail = None
        self.__size -= 1
        return val

class EmptyError(Exception):
    """Exception class for empty queue ADT"""
    pass

File: csharp-dsa/DSA/Queue/LinkedQueue.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace DSA.Queue
{
    public class LinkedQueue<T> : IQueue<T>
    {
        #region Constructor               

        public LinkedQueue()
        {
            Head = null;
            Tail = null;
            Size = 0;
        }

        #endregion

        #region Properties

        private Node<T> Head { get; set; }

        private Node<T> Tail { get; set; }

        public int Size { get; private set; }

        public bool IsEmpty => (Size == 0 && Head == null && Tail == null);

        public T Front
        {
            get
            {
                if (IsEmpty)
                    throw new QueueEmptyException("The queue is empty");
                return Head.Element;
            }
        }

        #endregion

        #region Methods

        public void Enqueue(T element)
        {
            var node = new Node<T>(null, element);
            if (IsEmpty)
                Head = node;                           
            else            
                Tail.Next = node;

            Tail = node;
            Size++;
        }

        public T Dequeue()
        {
            if (IsEmpty)
                throw new QueueEmptyException("The queue is empty");

            var val = Head.Element;
            if (Size == 1)                            
                Tail = null;
            Head = Head.Next;
            Size--;

            return val;
        }

        #endregion
    }    

    internal class Node<T>
    {
        internal Node<T> Next { get; set; }

        internal T Element { get; }

        internal Node(Node<T> next, T element)
        {
            Next = next;
            Element = element;
        }
    }
}
